### PR TITLE
Support to lock GPUs for each task according to GPU_PER_TASK environment

### PR DESCRIPTION
**Pull request by kikakkz**
original implementation lock bellman.gpu.lock to lock all GPUs in default. And in my test with cuda
single 3080 for C2: 15m52s
double 3080 for C2: 12m36s
so i think it's better to let user to choice if they need to use all GPUs for one C2 task.
a environment named GPU_PER_TASK is introduced to let user to set.
if GPU_PER_TASK is not set, then we use all GPUs for one C2, just like current implementation;
if GPU_PER_TASK == 0, just like above;
if GPU_PER_TASK > 0, use GPU_PER_TASK GPUs (up to devices.len()) for one C2 task.

### CODE REVIEW

**Comment by vmx** [comment]
With the [`CUDA_VISIBLE_DEVICES`](https://developer.nvidia.com/blog/cuda-pro-tip-control-gpu-visibility-cuda_visible_devices/) environment variable you can specify per program which GPUs should be visible. This way you can specify exactly which GPUs you would run it. Would this work for your use case instead of this patch?

**Comment by kikakkz** [comment]
> With the [`CUDA_VISIBLE_DEVICES`](https://developer.nvidia.com/blog/cuda-pro-tip-control-gpu-visibility-cuda_visible_devices/) environment variable you can specify per program which GPUs should be visible. This way you can specify exactly which GPUs you would run it. Would this work for your use case instead of this patch?

Not exactly. With CUDA_VISIBLE_DEVICES the whole process can only see the specified device. So if we have a machine with multiple card, we have to run multiple processes. and it's even worse if we run PC1/PC2/C2 in one worker. We have to isolate memory, cpu threads, GPU device, nvme space for each process. That's really complicated for automatic deployment script. With this PR, those all will be set in one process automatically.

**Comment by vmx** [comment]
@kikakkz thanks for the explanation. To make sure I understand your use case correctly. What you want is basically to be able to say "x number of GPUs form one unit". So let's say you have 6 GPUs in a machine, you set `GPU_PER_TASK=2`, then you kind of have 3 units of GPUs which can be used independently, but C2 would still use 2 GPUs.

**Comment by vmx** [comment]
Unrelated to the code, but still relevant: @kikakkz would it be possible for you to sign into CircleCI? I know it sounds weird, but this ways the CI would then be triggered correctly.

**Comment by kikakkz** [comment]
> Unrelated to the code, but still relevant: @kikakkz would it be possible for you to sign into CircleCI? I know it sounds weird, but this ways the CI would then be triggered correctly.

i try, but i struggle to fail，hmm~ let me try again~

**Comment by vmx** [comment]
> i try, but i struggle to fail，hmm~ let me try again~

Clearing the cookies might help (I got that information from CircleCI support, once there was a similar issue).

**Comment by kikakkz** [comment]
> 

yes, exactly. actually in my test, dual GPUs for one C2 do not have so much promotion of performance compare to single GPU for one C2 task. so in my practice, i would like to use single GPU for one C2 task, and run multiple C2 tasks concurrently within one worker process. and of course if i have 6 GPUs, i can set GPU_PER_TASK to 2, then i can run 3 C2 tasks concurrently.

**Comment by kikakkz** [comment]
> @kikakkz thanks for the explanation. To make sure I understand your use case correctly. What you want is basically to be able to say "x number of GPUs form one unit". So let's say you have 6 GPUs in a machine, you set `GPU_PER_TASK=2`, then you kind of have 3 units of GPUs which can be used independently, but C2 would still use 2 GPUs.

yes, exactly. actually in my test, dual GPUs for one C2 do not have so much promotion of performance compare to single GPU for one C2 task. so in my practice, i would like to use single GPU for one C2 task, and run multiple C2 tasks concurrently within one worker process. and of course if i have 6 GPUs, i can set GPU_PER_TASK to 2, then i can run 3 C2 tasks concurrently.

**Comment by kikakkz** [comment]
>

**Comment by kikakkz** [comment]
> > i try, but i struggle to fail，hmm~ let me try again~
> 
> Clearing the cookies might help (I got that information from CircleCI support, once there was a similar issue).

trying now, 😄

File: src/gpu/locks.rs [hunk 1]
```diff
@@ -14,31 +15,57 @@ use crate::gpu::{CpuGpuMultiexpKernel, GpuName};
...
-        GPULock(f)
+pub struct GPULock<'a>(Vec<(File, &'a Device, PathBuf)>);
+
+impl GPULock<'_> {
+    pub fn lock() -> GPULock<'static> {
+        let devices = Device::all();
+        let mut locks = Vec::new();
+
+        let mut gpu_per_task = match std::env::var("GPU_PER_TASK") {
+            Ok(val) => val.parse::<usize>().expect("GPU_PER_TASK must be number!"),
```
Comment (vmx) [inline src/gpu/locks.rs:37]: Putting in a non-numeric value shouldn't lead to a panic. I would rather log a warning and fall back to locking all GPUs.

File: src/gpu/locks.rs [hunk 2]
```diff
@@ -14,31 +15,57 @@ use crate::gpu::{CpuGpuMultiexpKernel, GpuName};
...
+            Err(_) => devices.len(),
+        };
+        if gpu_per_task == 0 {
+            gpu_per_task = devices.len();
+        }
+
+        for device in devices {
+            let uid = device.unique_id();
+            let gpu_lock_file = tmp_path(GPU_LOCK_NAME, uid);
+            debug!("Acquiring GPU {:?} lock at {:?} ...", uid, &gpu_lock_file);
```
Comment (vmx) [inline src/gpu/locks.rs:47]: Two things: 
 - I think logging only the filename (and not the UUID) is enough as the filename contains the UUID anyway.
 -  I propose adding a counter which lock you've acquired, that could make debugging easier in case something goes wrong.
  
 So it could look something like:

```rust
 debug!("Acquiring GPU lock {}/{} at {:?} ...", index, gpu_per_task, &gpu_lock_file);
```

Where you would change the for loop to something like:

```rust
for (index, device) in devices.enmuerate() {
```

File: src/gpu/locks.rs [hunk 3]
```diff
@@ -14,31 +15,57 @@ use crate::gpu::{CpuGpuMultiexpKernel, GpuName};
...
+            let f = File::create(&gpu_lock_file)
+                .unwrap_or_else(|_| panic!("Cannot create GPU {:?} lock file at {:?}", uid, &gpu_lock_file));
+            if f.try_lock_exclusive().is_err() {
+                continue
+            }
+            debug!("GPU {:?} lock acquired!", uid);
```
Comment (vmx) [inline src/gpu/locks.rs:53]: As the release we only have the filename, I propose making the logging similar. So I'd change that to:

```rust
debug!("GPU lock acquired at {}", gpu_lock_file);

File: src/gpu/locks.rs [hunk 4]
```diff
@@ -14,31 +15,57 @@ use crate::gpu::{CpuGpuMultiexpKernel, GpuName};
...
+        GPULock(locks)
     }
 }
 impl Drop for GPULock {
     fn drop(&mut self) {
-        self.0.unlock().unwrap();
-        debug!("GPU lock released!");
+        for f in &self.0 {
+            f.0.unlock().unwrap();
+            debug!("GPU {:?} lock released!", f.2);
```
Comment (vmx) [inline src/gpu/locks.rs:67]: And the same for releasing:

```rust
debug!("GPU lock released from {}", f.2);
```

**Comment by kikakkz** [comment]
@vmx i revoke access, clear cache, clear cookie then re-login circle ci, but still fail. after i try to get fail issue, and try to get configuration file, i get the url like 'https://app.circleci.com/projects/github/filecoin-project/bellperson/config/?branchName=&pipelineNumber=1784' in which branchName is missed, then i just give branchName to be master i can get config.yml file of circleci. could you please help ?

File: src/gpu/locks.rs [hunk 5]
```diff
@@ -50,7 +77,7 @@ impl Drop for GPULock {
 pub(crate) struct PriorityLock(File);
 impl PriorityLock {
     pub fn lock() -> PriorityLock {
-        let priority_lock_file = tmp_path(PRIORITY_LOCK_NAME);
+        let priority_lock_file = tmp_path(PRIORITY_LOCK_NAME, UniqueId::try_from(PRIORITY_LOCK_UID).unwrap());
```
Comment (vmx) [inline src/gpu/locks.rs:80]: I propose making the `tmp_path()` taking an option for the UUID, so you would just pass in `None` here instead having a "fake" UUID without much benefits.

**Comment by kikakkz** [comment]
and seems i cannot rerun the failed one because it miss all info in that record. should i recreate this PR and close this one ?

**Comment by kikakkz** [comment]
submit one more empty line and seems it triggered~ let's wait, 😄

**Review by vmx — COMMENTED** [review]
I think generally it looks good, my review contains minor things.

I wonder inf `GPU_PER_TASK` is a good name, as there currently is no notion of "task" in bellperson. So perhaps `BELLPERSON_GPUS_PER_LOCK` is a better name? Feel free to come up with even better names.

**Comment by vmx** [comment]
CI still seems weird. When you work through the code review, you can try to create another PR and we'll see if that helps.

File: src/gpu/locks.rs [hunk 6]
```diff
@@ -14,31 +15,57 @@ use crate::gpu::{CpuGpuMultiexpKernel, GpuName};
...
-impl GPULock {
-    pub fn lock() -> GPULock {
-        let gpu_lock_file = tmp_path(GPU_LOCK_NAME);
-        debug!("Acquiring GPU lock at {:?} ...", &gpu_lock_file);
-        let f = File::create(&gpu_lock_file)
-            .unwrap_or_else(|_| panic!("Cannot create GPU lock file at {:?}", &gpu_lock_file));
-        f.lock_exclusive().unwrap();
-        debug!("GPU lock acquired!");
-        GPULock(f)
+pub struct GPULock<'a>(Vec<(File, &'a Device, PathBuf)>);
```
Comment (vmx) [inline src/gpu/locks.rs:29]: Would it make sense to create a struct instead of having a tuple with 3 elements?

File: src/gpu/locks.rs (hunk 1 above)
Comment (kikakkz) [reply src/gpu/locks.rs:37]: will use BELLPERSON_GPUS_PER_LOCK to follow the convention firstly, 😄. i'm a bad namer...

**Comment by kikakkz** [comment]
https://github.com/filecoin-project/bellperson/pull/300 create this one to test circle instead
